extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,315 @@
+
+  public static Cache create(final RedisCacheConfig config)
+  {
+    final Supplier<ResourceHolder<ShardedJedis>> clientSupplier;
+
+    if (config.getPoolSize() > 1) {
+      clientSupplier = new LoadBalancingPool<>(
+          config.getPoolSize(),
+          new Supplier<ShardedJedis>()
+          {
+            @Override
+            public ShardedJedis get()
+            {
+              return new ShardedJedis(config.getShardsInfo(), Hashing.MURMUR_HASH);
+            }
+          }
+      );
+    } else {
+      clientSupplier = Suppliers.<ResourceHolder<ShardedJedis>>ofInstance(
+          StupidResourceHolder.create(new ShardedJedis(config.getShardsInfo(), Hashing.MURMUR_HASH))
+      );
+    }
+
+    return new RedisCache(clientSupplier, config);
+  }
+
+  public RedisCache(Supplier<ResourceHolder<ShardedJedis>> supplier, RedisCacheConfig config)
+  {
+    this.supplier = supplier;
+    prefix = config.getPrefix();
+    expiration = config.getExpiration();
+    transcoder = new RedisTranscoder();
+  }
+
+  @Override
+  public byte[] get(NamedKey key)
+  {
+    byte[] result = null;
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+      final byte[] itemKey = CacheImplUtils.computeKeyHash(prefix, key).getBytes();
+      final byte[] serialized = jedis.get(itemKey);
+
+      if (serialized != null) {
+        result = deserialize(key, serialized);
+        hitCount.incrementAndGet();
+      }
+      else {
+        missCount.incrementAndGet();
+      }
+    }
+    catch (IOException ex) {
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisConnectionException ex) {
+      if (isTimeout(ex)) {
+        timeoutCount.incrementAndGet();
+      }
+
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisException | LZ4Exception ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error getting cache item");
+      return null;
+    }
+
+    return result;
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    final byte[] serialized;
+    try {
+      serialized = serialize(key, value);
+    }
+    catch (NullPointerException ex) {
+      throw Throwables.propagate(ex);
+    }
+
+    try (ResourceHolder<ShardedJedis> clientHolder = supplier.get()) {
+      final ShardedJedis jedis = clientHolder.get();
+
+      final byte[] itemKey = CacheImplUtils.computeKeyHash(prefix, key).getBytes();
+      Jedis shard = jedis.getShard(itemKey);
+
+      if (expiration > 0) {
+        shard.psetex(itemKey, expiration, serialized);
+      }
+      else {
+        shard.set(itemKey, serialized);
+      }
+
+      final byte[] setKey = CacheImplUtils.computeNamespaceHash(prefix, key.namespace).getBytes();
+      shard = jedis.getShard(setKey);
+      shard.sadd(setKey, key.key);
+    }
+    catch (IOException | JedisConnectionException e) {
+      Throwables.propagate(e);
```

Comment:
Cache put operations should be fire and forget. We should not propagate errors and instead return to let the query complete. We can log errors but not prevent query results from being returned.